### PR TITLE
AdminHomePageUiTest: enhance to cover content of the sample course #7479

### LINKED ISSUE #7479: AdminHomePageUiTest: enhance to cover content of the sample course
As discussed in #7441

### PR DESCRIPTION
**Pull request by sujeet14108**
Fixes #7479

### CODE REVIEW

**Comment by sujeet14108** [comment]
@chowyb Please have a look
For feedback result now i have first expanded all panels and then check the html. Also for `Publish` and `Unpublish` now we are not checking the full HTML

**Comment by sujeet14108** [comment]
My Tests are passing locally so why is it failing on Travis ?

**Comment by chowyb** [comment]
@sujeet14108 I have updated your branch with the latest version of `master`. You should run GodMode on the test case after running `npm run build` again.

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java [hunk 1]
```diff
@@ -212,13 +213,39 @@ private void testCreateInstructorAction() throws Exception {
...
-        ______TS("new instructor can edit feedbackSession of sample course");
+        ______TS("new instructor can view result of First team feedback session of sample course");
+        coursesPage.loadInstructorHomeTab();
+        instructorHomePage = AppPage.getNewPageInstance(browser, InstructorHomePage.class);
+        resultsPage = instructorHomePage.clickFeedbackSessionViewResultsLink(demoCourseId, "First team feedback session");
+        if (resultsPage.isElementPresent("collapse-panels-button")) {
+            resultsPage.clickCollapseExpandButton();
+            resultsPage.waitForPanelsToExpand();
+        }
+        instructorHomePage.verifyHtmlMainContent("/newlyJoinedInstructorFeedbackFirstfeedbacksessionResultsPage.html");
```
Comment (chowyb) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: Should use `FirstFeedbackSession` instead of `Firstfeedbacksession`

File: src/main/java/teammates/common/util/Const.java [hunk 2]
```diff
@@ -1417,6 +1417,10 @@ private Const() {
                 + "<a href=\"/instructorHelp.jsp#gs\" target=\"_blank\">Getting Started Guide</a>.<br>A video tour"
                 + " is also available in our <a href=\"/index.jsp\" target=\"_blank\">home page</a>.";
 
+        public static final String FOR_NEW_INSTRUCTOR = "New to TEAMMATES? You may wish to have a look at our "
```
Comment (chowyb) [inline src/main/java/teammates/common/util/Const.java]: This might need renaming to distinguish it from `HINT_FOR_NEW_INSTRUCTOR`

**Comment by sujeet14108** [comment]
Ready for review :)

File: src/test/resources/pages/newlyJoinedInstructorFeedbackSessionSuccessEdited.html [hunk 3]
```diff
@@ -213,7 +213,7 @@ <h5 class="col-sm-2 margin-top-0">
                 <div class="col-sm-10">
                   <div class="panel panel-default panel-body mce-content-body content-editor" id="instructions" spellcheck="false">
                     <p>
-                      updated instructions
+                      Please give your feedback based on the following questions.
```
Comment (chowyb) [inline src/test/resources/pages/newlyJoinedInstructorFeedbackSessionSuccessEdited.html]: Not sure why this change is here.

Also, just wondering, is this HTML page still being used?

Comment (sujeet14108) [inline src/test/resources/pages/newlyJoinedInstructorFeedbackSessionSuccessEdited.html]: No, html is not used now. In fact we can now remove this file as it is not used.

Comment (chowyb) [inline src/test/resources/pages/newlyJoinedInstructorFeedbackSessionSuccessEdited.html]: Yep, if the file is no longer being used, it should be removed.

**Comment by sujeet14108** [comment]
@chowyb Can you look at why test is failing?

**Comment by chowyb** [comment]
@sujeet14108 this was due to a change in the CI environment, which has been addressed in a recent PR. Merging the latest version of `master` into the branch should resolve it.

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java [hunk 4]
```diff
@@ -212,49 +213,186 @@ private void testCreateInstructorAction() throws Exception {
...
-        ______TS("new instructor can click submit button of sample feedbackSession");
+        ______TS("new instructor can click submit button of First team feedback session but cannot submit");
+        instructorHomePage.loadInstructorHomeTab();
+        FeedbackSubmitPage fbsp = instructorHomePage.clickFeedbackSessionSubmitLink(
+                demoCourseId, "First team feedback session");
+        fbsp.verifyStatus(
+                "The feedback session is currently not open for submissions."
+                + " You can view the questions and any submitted responses for"
+                + " this feedback session but cannot submit new responses.");
+        //fbsp.verifyHtmlMainContent("/newlyJoinedInstructorFirstFeedbackSessionSubmissionEditPage.html");
```
Comment (chowyb) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: Remove this commented out code too. Check if there are any others.

Comment (sujeet14108) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: Done :)

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java [hunk 5]
```diff
@@ -212,49 +213,184 @@ private void testCreateInstructorAction() throws Exception {
...
         coursesPage.loadInstructorHomeTab();
         instructorHomePage = AppPage.getNewPageInstance(browser, InstructorHomePage.class);
-        instructorHomePage.clickFeedbackSessionViewResultsLink(demoCourseId, "Second team feedback session");
-        instructorHomePage.verifyHtmlMainContent("/newlyJoinedInstructorFeedbackResultsPage.html");
+        InstructorFeedbackResultsPage resultsPage =
+                instructorHomePage.clickFeedbackSessionViewResultsLink(demoCourseId, "First team feedback session");
+        if (resultsPage.isElementPresent("collapse-panels-button")) {
+            resultsPage.clickCollapseExpandButton();
+            resultsPage.waitForPanelsToExpand();
+        }
```
Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: The if-block repeats many times. Can we extract it as a method so that it can read like this?
`resultPage.exppandPanels()`

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java [hunk 6]
```diff
@@ -212,49 +213,184 @@ private void testCreateInstructorAction() throws Exception {
...
+        instructorHomePage.verifyHtmlMainContent("/newlyJoinedInstructorFirstFeedbackSessionResultsPage.html");
```
Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: Should this be `resultsPage.verify...`?

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java [hunk 7]
```diff
@@ -212,49 +213,184 @@ private void testCreateInstructorAction() throws Exception {
...
+        ______TS("new instructor can view result of Third team feedback session of sample course");
+        coursesPage.loadInstructorHomeTab();
+        instructorHomePage = AppPage.getNewPageInstance(browser, InstructorHomePage.class);
+        resultsPage = instructorHomePage.clickFeedbackSessionViewResultsLink(
+                demoCourseId, "Session with different question types");
+        if (resultsPage.isElementPresent("collapse-panels-button")) {
+            resultsPage.clickCollapseExpandButton();
+            resultsPage.waitForPanelsToExpand();
+        }
+        instructorHomePage.verifyHtmlMainContent("/newlyJoinedInstructorThirdFeedbackSessionResultsPage.html");
```
Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: This whole block 239-247 is repeated multiple times. May be extract into something like this?
```java
verifySessionResults(demoCourseId, "Session with different question types",
        "/newlyJoinedInstructorThirdFeedbackSessionResultsPage.html");
```

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java [hunk 8]
```diff
@@ -212,49 +213,184 @@ private void testCreateInstructorAction() throws Exception {
...
+        feedbackEditPage.clickEditSessionButton();
+
+        feedbackSession = getFeedbackSessionWithRetry(demoCourseId, "Session with different question types");
+        feedbackEditPage.editFeedbackSession(
+                feedbackSession.getStartTime(), feedbackSession.getEndTime(),
+                new Text("updated instructions"), feedbackSession.getGracePeriod());
         feedbackEditPage.reloadPage();
-        instructorHomePage.verifyHtmlMainContentWithReloadRetry("/newlyJoinedInstructorFeedbackSessionSuccessEdited.html");
+        feedbackSession = getFeedbackSessionWithRetry(demoCourseId, "Session with different question types");
+        assertEquals("<p>updated instructions</p>", feedbackSession.getInstructionsString());
```
Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: There is no need to check if instructor can _edit_ each session. Rather, we are checking to see if the data of the sample course is as expected.

**Comment by sujeet14108** [comment]
@damithc Please have a look.
Removed unnecessary test.

**Comment by damithc** [comment]
Given below are the parts that should remain. Stricken out parts can be removed because if the instructor has the right privileges and the course has the right data, there is no need to check if the instructor can do things like edit session because those functionalities are tested elsewhere for _all_ instructors, not just new instructors:
@chowyb what do you think? Am I missing something?

---

testCreateInstructorAction()

______TS("action fail & success: add multiple instructors");
______TS("action success: displayed instructor details are properly HTML-encoded");
______TS("action success : create instructor account and the account is created successfully after user's verification");
______TS("new instructor can see sample course in homepage");
~~______TS("new instructor can access sample coure enroll page");~~
______TS("new instructor can access sample coure details page");  **'course' misspelled**
______TS("new instructor can access sample coure edit page");  **this test verifies that the instructor has the right privileges**
~~______TS("new instructor can access sample coure feedback session adding page");~~
______TS("new instructor can view result of First team feedback session of sample course");
______TS("new instructor can view result of Second team feedback session of sample course");
______TS("new instructor can view result of Third team feedback session of sample course");
~~______TS("new instructor can edit First team feedback session of sample course");~~
~~______TS("new instructor can edit Second team feedback session of sample course");~~
~~______TS("new instructor can edit Third feedback session of sample course");~~
~~______TS("new instructor can click submit button of First team feedback session but cannot submit");~~
~~______TS("new instructor can click submit button of Second team feedback session");~~
~~______TS("new instructor can click submit button of Third feedback session but cannot submit");~~
______TS("action failure : invalid parameter"); **not sure what's being tested here though**
______TS("action success: course is accessible for newly joined instructor as student");

**Comment by chowyb** [comment]
@damithc the suggested changes looks fine to me. For the `action failure : invalid parameter` test, it appears to be testing for whether the provided email is invalid (in this case, an email address with no `@` symbol), and should likely be kept as a test.

**Comment by damithc** [comment]
> For the action failure : invalid parameter test, it appears to be testing for whether the provided email is invalid (in this case, an email address with no @ symbol), and should likely be kept as a test.

Thank for checking @chowyb 
Yes, we should keep it but may be document it better.

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java (hunk 5 above)
Comment (whipermr5) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: @sujeet14108 Still pending this change :)

**Comment by sujeet14108** [comment]
Ready for review :)

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java (hunk 5 above)
Comment (sujeet14108) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: Done :)

**Comment by sujeet14108** [comment]
@chowyb Done :)

**Comment by sujeet14108** [comment]
Ready for review :)

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java [hunk 9]
```diff
@@ -191,73 +184,35 @@ private void testCreateInstructorAction() throws Exception {
...
         InstructorCoursesPage coursesPage = AppPage.getNewPageInstance(browser, url, InstructorCoursesPage.class);
         coursesPage.waitForAjaxLoadCoursesSuccess();
-        coursesPage.unarchiveCourse(demoCourseId);
-        coursesPage.verifyHtmlMainContent("/newlyJoinedInstructorCoursesPageSampleCourseUnarhived.html");
+        coursesPage.loadInstructorHomeTab();
+        instructorHomePage = AppPage.getNewPageInstance(browser, InstructorHomePage.class);
+        InstructorFeedbackResultsPage resultsPage =
+                instructorHomePage.clickFeedbackSessionViewResultsLink(demoCourseId, "First team feedback session");
+        resultsPage.expandPanels();
+        resultsPage.verifyHtmlMainContent("/newlyJoinedInstructorFirstFeedbackSessionResultsPage.html");
```
Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: This block of code is repeated several times. Can be extracted as a method.

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java [hunk 10]
```diff
@@ -191,73 +184,35 @@ private void testCreateInstructorAction() throws Exception {
...
-        instructorHomePage.verifyHtmlMainContent("/newlyJoinedInstructorFeedbackSessionPublished.html");
-
-        feedbacksPage.logout();
+        resultsPage = instructorHomePage.clickFeedbackSessionViewResultsLink(
+                demoCourseId, "Session with different question types");
+        resultsPage.expandPanels();
+        resultsPage.verifyHtmlMainContent("/newlyJoinedInstructorThirdFeedbackSessionResultsPage.html");
 
+        //verify the given email is valid or not
         ______TS("action failure : invalid parameter");
```
Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: Instead of adding a comment, modify line 216 which is supposed to serve as a description of the test case.

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java [hunk 11]
```diff
@@ -177,11 +173,8 @@ private void testCreateInstructorAction() throws Exception {
         InstructorHomePage instructorHomePage = AppPage.getNewPageInstance(browser, InstructorHomePage.class);
         instructorHomePage.verifyHtmlMainContent("/newlyJoinedInstructorHomePage.html");
 
-        ______TS("new instructor can access sample coure enroll page");
-        InstructorCourseEnrollPage enrollPage = instructorHomePage.clickCourseEnrollLink(demoCourseId);
-        enrollPage.verifyHtmlMainContent("/newlyJoinedInstructorCourseEnrollPage.html");
-
         ______TS("new instructor can access sample coure details page");
```
Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: coure --> course

same typo in line 182

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java (hunk 9 above)
Comment (sujeet14108) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: Done :)

**Comment by sujeet14108** [comment]
Ready for review :)

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java [hunk 12]
```diff
@@ -191,73 +184,35 @@ private void testCreateInstructorAction() throws Exception {
...
-        instructorHomePage.loadInstructorHomeTab();
-        instructorHomePage.clickFeedbackSessionUnpublishLink(demoCourseId, "Second team feedback session");
-        instructorHomePage.verifyHtmlMainContent("/newlyJoinedInstructorFeedbackSessionUnpublished.html");
-
-        ______TS("new instructor can publish feedbackSession of sample course");
-        instructorHomePage.loadInstructorHomeTab();
-        instructorHomePage.clickFeedbackSessionPublishLink(demoCourseId, "Second team feedback session");
-        instructorHomePage.verifyHtmlMainContent("/newlyJoinedInstructorFeedbackSessionPublished.html");
-
-        feedbacksPage.logout();
```
Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: It's better now, but can be even more concise. 
This whole block
```java
        ______TS("new instructor can view result of First team feedback session of sample course");
        AppUrl url = createUrl(Const.ActionURIs.INSTRUCTOR_COURSES_PAGE)
                .withUserId(TestProperties.TEST_INSTRUCTOR_ACCOUNT);
        InstructorCoursesPage coursesPage = AppPage.getNewPageInstance(browser, url, InstructorCoursesPage.class);
        coursesPage.waitForAjaxLoadCoursesSuccess();
        coursesPage.loadInstructorHomeTab();
        instructorHomePage = AppPage.getNewPageInstance(browser, InstructorHomePage.class);
        InstructorFeedbackResultsPage resultsPage =
                instructorHomePage.clickFeedbackSessionViewResultsLink(demoCourseId, "First team feedback session");
        resultsPage.expandPanels();
        resultsPage.verifyHtmlMainContent("/newlyJoinedInstructorFirstFeedbackSessionResultsPage.html");

        ______TS("new instructor can view result of Second team feedback session of sample course");
        coursesPage.loadInstructorHomeTab();
        instructorHomePage = AppPage.getNewPageInstance(browser, InstructorHomePage.class);
        resultsPage =
                instructorHomePage.clickFeedbackSessionViewResultsLink(demoCourseId, "Second team feedback session");
        resultsPage.expandPanels();
        resultsPage.verifyHtmlMainContent("/newlyJoinedInstructorSecondFeedbackSessionResultsPage.html");

        ______TS("new instructor can view result of Third team feedback session of sample course");
        coursesPage.loadInstructorHomeTab();
        instructorHomePage = AppPage.getNewPageInstance(browser, InstructorHomePage.class);
        resultsPage = instructorHomePage.clickFeedbackSessionViewResultsLink(
                demoCourseId, "Session with different question types");
        resultsPage.expandPanels();
        resultsPage.verifyHtmlMainContent("/newlyJoinedInstructorThirdFeedbackSessionResultsPage.html");
```
can be collapsed to
```java
        ______TS("new instructor can view feedback sessions of sample course");
        AppUrl url = createUrl(Const.ActionURIs.INSTRUCTOR_COURSES_PAGE)
                .withUserId(TestProperties.TEST_INSTRUCTOR_ACCOUNT);

        verifyCanViewSesssion(demoCourseId, "First team feedback session",
                "/newlyJoinedInstructorFirstFeedbackSessionResultsPage.html");        
        verifyCanViewSesssion(demoCourseId, "Second team feedback session",
                "/newlyJoinedInstructorSecondFeedbackSessionResultsPage.html");  
        verifyCanViewSesssion(demoCourseId, "Third team feedback session",
                "/newlyJoinedInstructorThirdFeedbackSessionResultsPage.html");  
```

What do you think?

Comment (sujeet14108) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: Yes, this looks more concise and better :+1:

**Comment by sujeet14108** [comment]
@damithc have a look :)

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java [hunk 13]
```diff
@@ -115,6 +115,13 @@ public void clickCollapseExpandButton() {
         click(collapseExpandButton);
     }
 
+    public void expandPanels() {
+        if (isElementPresent("collapse-panels-button")) {
+            clickCollapseExpandButton();
+            waitForPanelsToExpand();
+        }
+    }
```
Comment (damithc) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java]: This method might be useful in other UI tests involving results pages. Of course that should be a separate PR. Just saying.

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java [hunk 14]
```diff
@@ -177,88 +173,31 @@ private void testCreateInstructorAction() throws Exception {
...
-
-        ______TS("action failure : invalid parameter");
+        ______TS("action failure : verify the given email is valid or not");
```
Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: This is more informative?
`______TS("action failure : trying to create instructor with an invalid email");`

**Comment by sujeet14108** [comment]
Done :)

File: src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java (hunk 14 above)
Comment (sujeet14108) [inline src/test/java/teammates/test/cases/browsertests/AdminHomePageUiTest.java]: Done

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java (hunk 13 above)
Comment (sujeet14108) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java]: In `InstructorFeedbackResultsPageUiTest`  there is similar method `clickCollapseExpandButtonAndWaitForPanelsToExpand` . We can now remove that method and to use this method ? 

There are other methods like `clickAjaxLoadedPanelAndWaitForExpansion` and  `clickCollapseExpandButtonAndWaitForPanelsToCollapse` in that class. Shouldn't they be in `InstructorFeedbackResultsPage` instead of `InstructorFeedbackResultsPageUiTest` ?

Comment (chowyb) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java]: Yep, from looking at the code, that would be a better approach.

**Comment by sujeet14108** [comment]
Done :)

**Comment by sujeet14108** [comment]
@chowyb Ready for review

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java (hunk 13 above)
Comment (sujeet14108) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java]: @damithc Any thought on this ?

Comment (damithc) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java]: > @damithc Any thought on this ?

Create a new issue ans tackle as a separate PR?
